cirq-google/cirq_google/engine/engine_client.py

```diff
@@ -95,6 +98,149 @@ def instance(cls):
+        self._grpc_client = grpc_client
+        self._request_queue = asyncio.Queue()
+        self._response_demux = ResponseDemux()
+        self._manage_stream_loop_future: Optional[duet.AwaitableFuture] = None
+
+    @property
+    def _executor(self) -> AsyncioExecutor:
+        # We must re-use a single Executor due to multi-threading issues in gRPC
+        # clients: https://github.com/grpc/grpc/issues/25364.
+        return AsyncioExecutor.instance()
+
+    @property
+    async def _request_iterator(self) -> AsyncIterator[quantum.QuantumRunStreamRequest]:
+        while not self._request_queue.empty():
+            yield await self._request_queue.get()
+
+    async def _manage_stream(self):
+        """Keeps the stream alive and routes responses to the appropriate request handler"""
+        while True:
+            try:
+                # TODO specify stream timeout below with exponential backoff
+                response_iterable = await self._grpc_client.quantum_run_stream(
+                    self._request_iterator
+                )
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except BaseException as e:
+                # TODO send halfclose to the stream upon CancelledError? How does that work?
+                self._response_demux.publish_exception(e)  # Raise to all request tasks
+                self._request_queue = asyncio.Queue()  # Clear requests
+
+    async def _make_request(
+        self, project_name: str, program: quantum.QuantumProgram, job: quantum.QuantumJob
+    ) -> quantum.QuantumResult:
+        """This method is executed in a separate asyncio Task for each request."""
+        current_request = quantum.QuantumRunStreamRequest(
+            parent=project_name,
+            create_quantum_program_and_job=quantum.CreateQuantumProgramAndJobRequest(
+                parent=project_name, quantum_program=program, quantum_job=job
+            ),
+        )
+        get_result_request = quantum.QuantumRunStreamRequest(
+            parent=project_name, get_quantum_result=quantum.GetQuantumResultRequest(parent=job.name)
+        )
+
+        response_future: Optional[asyncio.Future] = None
+        response: Optional[quantum.QuantumRunStreamResponse] = None
+        while response is None:
+            try:
+                response_future = self._response_demux.subscribe(current_request)
+                await self._request_queue.put(current_request)
+                response = await response_future
+                if response.result is not None:
+                    return response.result
+
+            except GoogleAPICallError:
+                # TODO how to distinguish between program not found vs job not found?
+                # TODO handle QuantumJob response and retryable StreamError.
+                # TODO Send a CreateProgramAndJobRequest or CreateJobRequest if either program or
```

Comment:
Not sure what this means. e.g., a "job doesn't exist" should only happen during a `GetQuantumResultRequest`, but that value won't have enough context to create a new job.

Reply:
`GetQuantumResultRequest` is only sent when a previous `CreateProgramAndJobRequest` failed for some reason (most likely a stream failure) within this asyncio task, so this task does have the full program and job context to recreate the job.

Getting a "job doesn't exist" means either the "create program" part or the "create job" part failed. We could try recreating just the job first by sending a `CreateJobRequest`, and if that fails with a "program doesn't exist" error, we can try the full `CreateProgramAndJobRequest` again.